### ISSUE TITLE
Test failures on PowerPC arch

### ISSUE THREAD

**Issue by sagar0** [facebook/rocksdb#3347]
@adamretter alerted us that `make check` is failing on PPC64le (powerpc 64-bit little-endian) architecture on RocksDB master HEAD. On investigating further, I noticed a couple of tests have been failing since RocksDB 5.7 release. 
The two specific tests that are failing are: 
- FullBloomTest.FullVaryingLengths
- DBPropertiesTest.AggregatedTablePropertiesAtLevel

I narrowed the issue down to be due to ccf5f08f882038e8b481fafa0a0c0b1a04f6a390 introduced in #2579 . The tests work without this commit and fail with this commit included. 

### Steps to reproduce the behavior
Either `make check`
or
`TMP_DIR=/dev/shm ./bloom_test --gtest_filter=FullBloomTest.FullVaryingLengths`
`TMP_DIR=/dev/shm ./db_properties_test --gtest_filter=DBPropertiesTest.AggregatedTablePropertiesAtLevel` and

```
ubuntu@hortonworks-2:/tmp/rdb/rocksdb$ make bloom_test db_properties_test -j8
...
...
ubuntu@hortonworks-2:/tmp/rdb/rocksdb$ TMP_DIR=/dev/shm ./bloom_test --gtest_filter=FullBloomTest.FullVaryingLengths
Note: Google Test filter = FullBloomTest.FullVaryingLengths
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from FullBloomTest
[ RUN      ] FullBloomTest.FullVaryingLengths
False positives:  0.00% @ length =      1 ; bytes =    133
False positives:  0.00% @ length =      2 ; bytes =    133
False positives:  0.00% @ length =      3 ; bytes =    133
False positives:  0.00% @ length =      4 ; bytes =    133
False positives:  0.00% @ length =      5 ; bytes =    133
False positives:  0.00% @ length =      6 ; bytes =    133
False positives:  0.00% @ length =      7 ; bytes =    133
False positives:  0.00% @ length =      8 ; bytes =    133
False positives:  0.00% @ length =      9 ; bytes =    133
False positives:  0.00% @ length =     10 ; bytes =    133
False positives:  0.00% @ length =     20 ; bytes =    133
False positives:  0.00% @ length =     30 ; bytes =    133
False positives:  0.02% @ length =     40 ; bytes =    133
False positives:  0.04% @ length =     50 ; bytes =    133
False positives:  0.15% @ length =     60 ; bytes =    133
False positives:  0.27% @ length =     70 ; bytes =    133
False positives:  0.59% @ length =     80 ; bytes =    133
False positives:  0.83% @ length =     90 ; bytes =    133
False positives:  1.17% @ length =    100 ; bytes =    133
util/bloom_test.cc:287: Failure
Expected: (FilterSize()) <= ((size_t)((length * 10 / 8) + 128 + 5)), actual: 389 vs 383
200
[  FAILED  ] FullBloomTest.FullVaryingLengths (8 ms)
[----------] 1 test from FullBloomTest (8 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (8 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] FullBloomTest.FullVaryingLengths

 1 FAILED TEST


ubuntu@hortonworks-2:/tmp/rdb/rocksdb$ TMP_DIR=/dev/shm ./db_properties_test --gtest_filter=DBPropertiesTest.AggregatedTablePropertiesAtLevel
Note: Google Test filter = DBPropertiesTest.AggregatedTablePropertiesAtLevel
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DBPropertiesTest
[ RUN      ] DBPropertiesTest.AggregatedTablePropertiesAtLevel
db/db_properties_test.cc:200: Failure
Expected: (static_cast<double>(dbl_a - dbl_b) / (dbl_a + dbl_b)) < (bias), actual: 0.599198 vs 0.5
...
```

Arch and OS details the machine I have access to shows:
```
ubuntu@hortonworks-2:/tmp/rdb/rocksdb$ uname -a
Linux hortonworks-2 4.4.0-101-generic #124-Ubuntu SMP Fri Nov 10 18:29:11 UTC 2017 ppc64le ppc64le ppc64le GNU/Linux
ubuntu@hortonworks-2:/tmp/rdb/rocksdb$
ubuntu@hortonworks-2:/tmp/rdb/rocksdb$ cat /proc/cpuinfo
processor       : 0
cpu             : POWER8E (raw), altivec supported
clock           : 3425.000000MHz
revision        : 2.1 (pvr 004b 0201)
```

**Comment by sagar0** [comment]
@grooverdan I was wondering if you could help us in understanding more and in fixing this issue related to Cache line changes, as we don't know much about powerpc arch.

**Comment by grooverdan** [comment]
sure I'll trace it down. Sorry I lost track of it failing.

**Comment by grooverdan** [comment]
WIP:

Looking at P8 with the cache-line size patch reverted. The expansion of the FilterSize() (bytes) occurs once length approaches it. In the above FilterSize() remains at 133 as the length doesn't approach it however it exceeds the 2% margin to assert. The false positive rate below reduces as the FilterSize() increases.

<pre>FullBloomTest.FullVaryingLengths
[ RUN      ] FullBloomTest.FullVaryingLengths
False positives:  0.00% @ length =      1 ; bytes =     69
False positives:  0.00% @ length =      2 ; bytes =     69
False positives:  0.01% @ length =      3 ; bytes =     69
False positives:  0.02% @ length =      4 ; bytes =     69
False positives:  0.02% @ length =      5 ; bytes =     69
False positives:  0.02% @ length =      6 ; bytes =     69
False positives:  0.05% @ length =      7 ; bytes =     69
False positives:  0.06% @ length =      8 ; bytes =     69
False positives:  0.07% @ length =      9 ; bytes =     69
False positives:  0.08% @ length =     10 ; bytes =     69
False positives:  0.10% @ length =     20 ; bytes =     69
False positives:  0.23% @ length =     30 ; bytes =     69
False positives:  0.57% @ length =     40 ; bytes =     69
False positives:  0.95% @ length =     50 ; bytes =     69
False positives:  0.08% @ length =     60 ; bytes =    197
False positives:  0.12% @ length =     70 ; bytes =    197
False positives:  0.19% @ length =     80 ; bytes =    197
False positives:  0.26% @ length =     90 ; bytes =    197
False positives:  0.31% @ length =    100 ; bytes =    197
...
</pre>

DBPropertiesTest.AggregatedTablePropertiesAtLevel:

With some debugging before the assert:
<pre>
data_size: base 468380 new_tp: 458000
index_size: base  11405 new_tp:   7686
filter_size: base   7182 new_tp:   2500
num_data_blocks: base    526 new_tp:    427
num_data_blocks: base    526 new_tp:    427
[  FAILED  ] DBPropertiesTest.AggregatedTablePropertiesAtLevel (4905 ms)
</pre>

With patch reverted - significantly smaller filter_size for the rest of the values being the same.
<pre>
data_size: base 468380 new_tp: 458000
index_size: base  11405 new_tp:   7686
filter_size: base   3726 new_tp:   2500
num_data_blocks: base    526 new_tp:    427
</pre>

**Comment by sagar0** [comment]
@grooverdan - Thanks for investing further and finding that this could happen on x86_64 too if CACHE_LINE_SIZE changes to 128, due to a potential FilterSize bug. 
Since CACHE_LINE_SIZE of 128 is used only on powerpc and arm architectures for now, can you confirm that 128 is the right choice? If so, we could move our efforts to investigating and fixing filter-size recalculation issues.

**Comment by grooverdan** [comment]
These cache sizes from  ccf5f08f882038e8b481fafa0a0c0b1a04f6a390 are correct for powerpc, arm, s360 and x86_64.

The patch I used for verbose db_properties is https://github.com/grooverdan/rocksdb/commit/36fd39bcb30040ec91604ebecd821e33e415bab3 (though its far too noisy to merge). Sorry I didn't get into working out the filter-size recalculation.

**Comment by adamretter** [comment]
@sagar0 @grooverdan I am afraid I am not sure of how this needs to be moved forward. Do I understand that this is a bug in the FilterSize calculation of RocksDB as opposed to a platform configuration issue?

**Comment by grooverdan** [comment]
Yes. There's nothing platform related here. Manually set x86_64 cacheline size to 128 and you'll see the same test failure. Something in FilterSize is implicitly assuming a 64 cache line size in its calculations.
